libraries/net/node.cpp

```diff
@@ -283,6 +193,91 @@ namespace graphene { namespace net { namespace detail {
 #define MAXIMUM_NUMBER_OF_BLOCKS_TO_HANDLE_AT_ONE_TIME 200
 #define MAXIMUM_NUMBER_OF_BLOCKS_TO_PREFETCH (10 * MAXIMUM_NUMBER_OF_BLOCKS_TO_HANDLE_AT_ONE_TIME)
 
+   /******
+    * Use information passed from command line or config file to advertise nodes
+    */
+   class list_address_builder : public node_impl::address_builder
+   {
+      public:
+      list_address_builder(fc::optional<std::vector<std::string>> address_list)
+      {
+         if (address_list.valid())
+         {
+            advertise_list.reserve( address_list->size() );
+            auto& list = advertise_list;
+            std::for_each( address_list->begin(), address_list->end(), [&list]( std::string str ) {
+                  // ignore fc exceptions (like poorly formatted endpoints)
+                  try
+                  {
+                     list.emplace_back( graphene::net::address_info(
+                        fc::ip::endpoint::from_string(str),
+                        fc::time_point_sec(),
+                        fc::microseconds(0),
+                        node_id_t(),
+                        peer_connection_direction::unknown,
+                        firewalled_state::unknown ));
+                  }
+                  catch(const fc::exception& ) {
+                     wlog( "Address ${addr} invalid.", ("addr", str) );
+                  } 
+               } );
+         }
+      }
+
+      void build(node_impl* impl, address_message& reply)
+      {
+         reply.addresses = advertise_list;
+      }
+      private:
+      std::vector<graphene::net::address_info> advertise_list;
+   };
+
+   /****
+    * Advertise all nodes except a predefined list
+    */
+   class exclude_address_builder : public node_impl::address_builder
+   {
+      public:
+      exclude_address_builder(const fc::optional<std::vector<std::string>>& address_list)
+      {
+         if (address_list.valid())
```

Comment:
IMO should treat is as an error if the list is absent or empty.

Reply:
Done